### PR TITLE
Feature/combine node with rest

### PR DESCRIPTION
**Pull request by adrianryt**

### CODE REVIEW

File: src/main/java/rest/NodeController.java [hunk 1]
```diff
@@ -1,16 +1,67 @@
...
+    @Autowired
+    private Node node;
+
+    @RequestMapping(value = "/connect", method = RequestMethod.POST)
+    public String connectHandler() {
+        // TODO handle this, for now static response
+        this.node.connectToServer();
+        this.node.startNode();
+        JsonObject response = new JsonObject();
+        response.addProperty("message", "OK");
```
Comment (Atheam) [inline src/main/java/rest/NodeController.java:28]: I think we should respond to user with node ID in this method. Either by waiting or pushing notification after some time.

File: src/main/java/rest/NodeController.java [hunk 2]
```diff
@@ -1,16 +1,67 @@
...
+        return response.toString();
+    }
+
+    @RequestMapping(value = "register", method = RequestMethod.POST)
+    public String registerHandler() {
```
Comment (Atheam) [inline src/main/java/rest/NodeController.java:33]: We should first check if node is connected and has ID assigned, also ID should be sent to user in /connect handler not here.

File: src/main/java/rest/NodeController.java [hunk 3]
```diff
@@ -1,16 +1,67 @@
...
+        this.node.registerNode();
+        JsonObject response = new JsonObject();
+        response.addProperty("message", "OK");
+        response.addProperty("ID", this.node.getID());
+        return response.toString();
+    }
+
+    @RequestMapping(value = "/broadcast", method = RequestMethod.POST)
+    public String broadcastHandler() {
```
Comment (Atheam) [inline src/main/java/rest/NodeController.java:42]: Check if node is connected and registered

File: src/main/java/node/Node.java [hunk 4]
```diff
@@ -9,82 +10,151 @@
...
+                }
+                System.out.println("RECEIVED DATA:" + new String(message.getData()));
+                break;
+            case REQUEST_BROADCAST:
+                this.requestBroadcast();
+                break;
+        }
+    }
+
+    public void addTransactionToMemPool(Transaction transaction) {
```
Comment (Atheam) [inline src/main/java/node/Node.java:137]: I think we should move this to some BlockchainHandler class to keep Node class functionality restricted to processing Messeges, same as we did with ServerSessionHandler and PeerConnectionHandler

File: src/main/java/node/Node.java [hunk 5]
```diff
@@ -9,82 +10,151 @@
...
+        this.memPool.addTransaction(transaction);
+        // TODO does it work?
+        synchronized (this.miner) {
+            this.miner.notify();
+        }
+    }
+
+    private void handleBlockchainFromOtherNode(List<Block> blockchain) {
```
Comment (Atheam) [inline src/main/java/node/Node.java:146]: Same as above, move to another class

File: src/main/java/node/Listeners/UdpListener.java [hunk 6]
```diff
@@ -30,8 +29,9 @@ private void listenForUdpPackets() throws IOException {
         DatagramPacket datagram = new DatagramPacket(new byte[1024], 1024);
         while (true) {
             socket.receive(datagram);
-            System.out.println("Received upd packet from: " + datagram.getAddress() + ":" + datagram.getPort() + " with data: " + new String(datagram.getData()));
-            Event event = new Event(datagram.getData());
+//            System.out.println("Received upd packet from: " + datagram.getAddress() + ":" + datagram.getPort() + " with data: " + new String(datagram.getData()));
+            System.out.println(new String(Arrays.copyOf(datagram.getData(), datagram.getLength())));
```
Comment (Atheam) [inline src/main/java/node/Listeners/UdpListener.java:33]: Move trimming bytes array to seperate metod.

File: src/main/java/blockchain/Blockchain.java [hunk 7]
```diff
@@ -1,19 +1,62 @@
...
+import java.util.Date;
+import java.util.LinkedList;
 import java.util.List;
 
 public class Blockchain {
     //TODO consider Thread safe queue and add DataBase
-    private List<Block> blockchain = new ArrayList<>();
+    public List<Block> blockchain = new ArrayList<>();
+    //TODO this eventListener maybe je**** method
+    private EventListener listener;
```
Comment (Atheam) [inline src/main/java/blockchain/Blockchain.java]: Make it a list of EventListeners (even if now we use only one listener) and also change subscribe and notify to handle a list of listeners. Also consider moving this event managing functionality to abstract class (look TcpListener or UdpListener).

**Review by Atheam — CHANGES_REQUESTED** [review]

File: src/main/java/rest/NodeController.java (hunk 1 above)
Comment (adrianryt) [reply src/main/java/rest/NodeController.java:28]: It is in TODO list

File: src/main/java/rest/NodeController.java (hunk 2 above)
Comment (adrianryt) [reply src/main/java/rest/NodeController.java:33]: It is in TODO list

File: src/main/java/rest/NodeController.java (hunk 3 above)
Comment (adrianryt) [reply src/main/java/rest/NodeController.java:42]: It is in TODO list

File: src/main/java/node/Listeners/UdpListener.java (hunk 6 above)
Comment (adrianryt) [reply src/main/java/node/Listeners/UdpListener.java:33]: You did this 😄

File: src/main/java/blockchain/Blockchain.java (hunk 7 above)
Comment (adrianryt) [reply src/main/java/blockchain/Blockchain.java]: read TODO

File: src/main/java/node/Listeners/UdpListener.java (hunk 6 above)
Comment (Atheam) [reply src/main/java/node/Listeners/UdpListener.java:33]: ?

File: src/main/java/blockchain/Blockchain.java (hunk 7 above)
Comment (Atheam) [reply src/main/java/blockchain/Blockchain.java]: i think its straight forward you can just change it no need to postpone it.

**Review by Atheam — APPROVED** [review]

**Review by wawel37 — APPROVED** [review]

**Pull request merged**